File: 205-isomorphic-strings/205-isomorphic-strings.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        sChars, tChars = {}, {}
        for i in range(len(s)):
            if s[i] in sChars:
                if sChars[s[i]] != t[i]:
                    return False
            elif t[i] in tChars:
                return False
            else:
                sChars[s[i]] = t[i]
                tChars[t[i]] = s[i]
                
        return True

File: 205-isomorphic-strings/205-isomorphic-strings.java
class Solution {
    private Map<Character, Character> sCharMap, tCharMap;
    
    public boolean isIsomorphic(String s, String t) {
        if (s.length() == 1) return true;
        
        sCharMap = new HashMap<>();
        tCharMap = new HashMap<>();
        for (int i = 0; i < s.length(); i++) {
            char sc = s.charAt(i), tc = t.charAt(i);
            if (!isMapped(sCharMap, sc) && !isMapped(tCharMap, tc)) {
                sCharMap.put(sc, tc);
                tCharMap.put(tc, sc);
            } else if (!isConnected(sc, tc)) {
                return false;
            }
        }
        return true;
    }
    
    private boolean isMapped(Map<Character, Character> charMap, char c) {
        return charMap.containsKey(c);
    }
    
    private boolean isConnected(char sChar, char tChar) {
        return Character.valueOf(tChar).equals(sCharMap.get(sChar)) 
            && Character.valueOf(sChar).equals(tCharMap.get(tChar));
    }
}